src/Interpreters/Cache/FileCache.cpp

```diff
@@ -170,6 +187,35 @@ void FileCache::assertInitialized() const
 }
 
 void FileCache::initialize()
+{
+    // Prevent initialize() from running twice. This may be caused by two cache disks being created with the same path (see integration/test_filesystem_cache).
+    if (initialize_called.exchange(true))
```

Comment:
All initialize() calls should wait if the first initialize() is still in progress, otherwise they will try to use cache and will get "Cache is not initialized" error

Reply:
I've substituted the guard using the atomic for a call to `callOnce`, so that all threads calling `initialize()` will wait until the first `initialize()` is finished if `load_metadata_asynchronously` is disabled.

In the case where it is enabled, I think all uses of the cache are guarded by `isInitialized()` calls, do you have some execution path in mind that could cause this error? I've also tried creating a lot of tables with the same cache and inserting in them concurrently with a sleep in `loadMetadata` and haven't been able to get that error.

Reply:
We call initialize() for cache when we start disks on server startup, all disks are started one by one synchronously, so currently intialize() should not be called in parallel unless some other thread during startup accesses the cache in the meantime (cannot remember such case now), but anyway we should be prepared for this to happen. std::call_once is ok.